spring-test/src/main/java/org/springframework/test/web/servlet/request/MockHttpServletRequestBuilder.java

```diff
@@ -696,11 +704,49 @@ private void updatePathRequestProperties(MockHttpServletRequest request, String
 						"Invalid servlet path [" + this.servletPath + "] for request URI [" + requestUri + "]");
 			}
 			String extraPath = requestUri.substring(this.contextPath.length() + this.servletPath.length());
-			this.pathInfo = (StringUtils.hasText(extraPath) ? extraPath : null);
+			this.pathInfo = (StringUtils.hasText(extraPath) ? decodeRequestString(request, extraPath) : null);
 		}
 		request.setPathInfo(this.pathInfo);
 	}
 
+	/**
+	 * Performs URL decoding on the provided source using the encoding from the request.
+	 *
+	 * @param request the request to use to determine the encoding
+	 * @param source the source to URL encode
+	 * @return the URL encoded string
+	 */
+	private String decodeRequestString(HttpServletRequest request, String source) {
+		String enc = determineEncoding(request);
+		try {
+			return UriUtils.decode(source, enc);
+		}
+		catch (UnsupportedEncodingException ex) {
+			if (logger.isWarnEnabled()) {
+				logger.warn("Could not decode request string [" + source + "] with encoding '" + enc +
+						"': falling back to platform default encoding; exception message: " + ex.getMessage());
+			}
+			return URLDecoder.decode(source);
+		}
+	}
+
+	/**
+	 * Determine the encoding for the given request.
+	 * Can be overridden in subclasses.
+	 * <p>The default implementation checks the request encoding,
+	 * falling back to {@code WebUtils.DEFAULT_CHARACTER_ENCODING}
+	 * @param request current HTTP request
+	 * @return the encoding for the request (never {@code null})
+	 * @see javax.servlet.ServletRequest#getCharacterEncoding()
+	 */
+	private String determineEncoding(HttpServletRequest request) {
+		String enc = request.getCharacterEncoding();
+		if (enc == null) {
+			enc = WebUtils.DEFAULT_CHARACTER_ENCODING;
+		}
+		return enc;
+	}
+
```

Comment:
`decodeRequestString` and `determineEncoding` appear identical to the same methods in `UrlPathHelper`. Why not just use `UrlPathHelper#decodeRequestString` instead?

Reply:
You are right. I copied it from there and somehow missed that was a public method.

Reply:
I updated the Pull Request with your suggestion.